scripting/get5/recording.sp

```diff
@@ -94,9 +94,71 @@ public Action Timer_FireStopRecordingEvent(Handle timer, DataPack pack) {
   Call_PushCell(event);
   Call_Finish();
   EventLogger_LogAndDeleteEvent(event);
+
+  UploadDemoToServer(demoFileName);
   return Plugin_Handled;
 }
 
+void UploadDemoToServer(const char[] demoFileName) {
+  char demoUrl[1024];
+  char demoAuthKey[1024];
+  char demoAuthValue[1024];
+
+  g_DemoUploadUrlCvar.GetString(demoUrl, sizeof(demoUrl));
+  g_DemoUploadAuthKeyCvar.GetString(demoAuthKey, sizeof(demoAuthKey));
+  g_DemoUploadAuthValueCvar.GetString(demoAuthValue, sizeof(demoAuthValue));
+
+  if (StrEqual(demoUrl, "")) {
+    LogDebug("Will not upload any demoes as upload URL is not set.");
+    return;
+  }
+
+  if (!LibraryExists("SteamWorks")) {
+    LogDebug("Cannot upload demoes to a web server without the SteamWorks extension running.");
+    return;
+  }
+
+  char formattedUrl[1024];
+  strcopy(formattedUrl, sizeof(formattedUrl), demoUrl);
+  ReplaceString(formattedUrl, sizeof(formattedUrl), "\"", "");
+
+  // Add the protocol strings if missing (only http).
+  if (StrContains(formattedUrl, "http://") == -1 && StrContains(formattedUrl, "https://") == -1) {
+    Format(formattedUrl, sizeof(formattedUrl), "http://%s", formattedUrl);
+  }
+
+  LogDebug("UploadDemoToServer: formattedUrl (SteamWorks) = %s", formattedUrl);
+  Handle demoRequest = SteamWorks_CreateHTTPRequest(k_EHTTPMethodPOST, formattedUrl);
+
+  if (demoRequest == INVALID_HANDLE) {
+    LogDebug("Failed to create request %s", formattedUrl);
+    return;
+  }
+
+  // Set the auth keys only if they are defined. If not, we can still technically POST
+  // to an end point that has no authentication.
+  if (!StrEqual(demoAuthKey, "") && !StrEqual(demoAuthValue, "")) {
+    if (!SteamWorks_SetHTTPRequestHeaderValue(demoRequest, demoAuthKey, demoAuthValue)) {
+      LogError("Failed to add auth key and token to call. Will end request here.");
+      delete demoRequest;
+      return;
+    }
+  }
+
+  if (!SteamWorks_SetHTTPRequestRawPostBodyFromFile(demoRequest, "application/octet-stream",
+                                                    demoFileName)) {
+    delete demoRequest;
+    return;
+  }
+  SteamWorks_SetHTTPCallbacks(demoRequest, DemoRequestCallback);
+  if (SteamWorks_SendHTTPRequest(demoRequest)) {
+    // Store the filename of the demo for the callback.
+    Format(g_DemoFileNameLastUploaded, sizeof(g_DemoFileNameLastUploaded), demoFileName);
+  } else {
+    LogError("UploadDemoToServer: An error occured when trying to send the HTTP request.");
```

Comment:
Do we need to delete `demoRequest` here too, or does the Callback take care of the deletion if the `SteamWorks_SendHTTPRequest` fails?

Reply:
You must delete it.

Reply:
I'm not sure that this **can** fail though, to be fair. I think it might only fail if there is a problem with the parameters.

Reply:
The SteamWorks code looks like this for the call, so it would appear to only fail if `demoRequest` is an invalid handle:

```c
ISteamHTTP *pHTTP;
SteamWorksHTTPRequest *pRequest = GetRequestPointer(pHTTP, pContext, params[1]);
if (pRequest == NULL)
{
	return 0;
}

SteamAPICall_t hCall;
cell_t result = pHTTP->SendHTTPRequest(pRequest->request, &hCall) ? 1 : 0;

SetCallbacks(hCall, pRequest);
return result;
```

 ... which at this point should not be possible.